File: Go/Intersection of Two Linked Lists.go
// 40ms, 86.49%; 8.6MB, 100.00%
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func getIntersectionNode(headA, headB *ListNode) *ListNode {
	len1, len2 := length(headA), length(headB)
	if len1 < len2 {
		headB = advance(headB, len2-len1)
	} else {
		headA = advance(headA, len1-len2)
	}
	for headA != nil && headA != headB {
		headA = headA.Next
		headB = headB.Next
	}
	return headA
}

func length(head *ListNode) int {
	ret := 0
	for head != nil {
		ret++
		head = head.Next
	}
	return ret
}

func advance(head *ListNode, dist int) *ListNode {
	for i := dist; i > 0; i-- {
		head = head.Next
	}
	return head
}

File: Java/Intersection of Two Linked Lists.java
// 1ms, 97.67%; 41.8MB, 5.71%
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public static ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int lenA = length(headA), lenB = length(headB);
        if (lenA < lenB) {
            headB = advance(headB, lenB - lenA);
        } else {
            headA = advance(headA, lenA - lenB);
        }
        while (headA != null) {
            if (headA == headB) {
                return headA;
            }
            headA = headA.next;
            headB = headB.next;
        }
        return null;
    }

    private static int length(ListNode node) {
        int ret = 0;
        while (node != null) {
            ++ret;
            node = node.next;
        }
        return ret;
    }

    private static ListNode advance(ListNode node, int x) {
        while (x-- > 0) {
            node = node.next;
        }
        return node;
    }
}
